vms/evm/signer/kms_signer.go

```diff
@@ -0,0 +1,176 @@
+	if err != nil {
+		return nil, err
+	}
+	eoa := crypto.PubkeyToAddress(*pubKey)
+
+	return &KMSSigner{
+		keyID:  keyID,
+		client: kmsClient,
+		pubKey: asn1pubk.PublicKey.Bytes,
+		eoa:    eoa,
+	}, nil
+}
+
+func (s *KMSSigner) SignTx(tx *types.Transaction, evmChainID *big.Int) (*types.Transaction, error) {
+	signer := types.LatestSignerForChainID(evmChainID)
+	h := signer.Hash(tx).Bytes()
+	signInput := kms.SignInput{
+		KeyId:            aws.String(s.keyID),
+		SigningAlgorithm: signingAlgorithm,
+		MessageType:      messageType,
+		Message:          h,
+	}
+
+	// Sign the hash of the transaction via KMS. The returned signature contains the R and S values in ASN.1 format
+	signOutput, err := s.client.Sign(context.Background(), &signInput)
+	if err != nil {
+		return nil, err
+	}
+	var sigAsn1 asn1ECSig
+	_, err = asn1.Unmarshal(signOutput.Signature, &sigAsn1)
+	if err != nil {
+		return nil, err
+	}
+
+	sigBytes, err := s.getEthereumSignature(h, sigAsn1.R.Bytes, sigAsn1.S.Bytes)
+	if err != nil {
+		return nil, err
+	}
+	return tx.WithSignature(signer, sigBytes)
+}
+
+func (s *KMSSigner) Address() common.Address {
+	return s.eoa
+}
+
+// Recover the EIP-155 signature from the KMS signature
+// KMS returns the signature in ASN.1 format, but the EIP-155 signature is in R || S || V format,
+// so we need to test both V = 0 and V = 1 against the recovered public key
+func (s *KMSSigner) getEthereumSignature(txHash []byte, rBytes []byte, sBytes []byte) ([]byte, error) {
+	rsSignature := append(adjustSignatureLength(rBytes), adjustSignatureLength(sBytes)...)
+	if len(rsSignature) != signatureLength {
+		return nil, errors.New("rs signature length is not 64 bytes")
+	}
+
+	// Check if the public key can be reconstructed with v=0
+	if signature, err := s.checkRSVSignature(txHash, rsSignature, 0); signature != nil {
+		return signature, nil
+	} else if err != nil {
+		return nil, err
+	}
+
+	// Check if the public key can be reconstructed with v=1
+	if signature, err := s.checkRSVSignature(txHash, rsSignature, 1); signature != nil {
+		return signature, nil
+	} else if err != nil {
+		return nil, err
+	}
+
+	return nil, errors.New("cannot reconstruct public key from sig")
+}
+
+// Checks if the public key can be reconstructed from the signature with the given v value
+// Returns a non-nil signature if the public key can be reconstructed, nil if it cannot
+func (s *KMSSigner) checkRSVSignature(txHash []byte, rsSignature []byte, v uint8) ([]byte, error) {
+	signature := append(rsSignature, []byte{v}...)
+	pubKey, err := crypto.Ecrecover(txHash, signature)
+	if err != nil {
+		return nil, err
+	}
+	if bytes.Equal(pubKey, s.pubKey) {
+		return signature, nil
+	}
+	return nil, nil
+}
+
+// Trim null bytes and pad with zeros to 32 bytes
+func adjustSignatureLength(buffer []byte) []byte {
+	if len(buffer) > 32 {
+		buffer = bytes.TrimLeft(buffer, "\x00")
+	}
+
+	// pad to 32 bytes
+	pad := 32 - len(buffer)
```

Comment:
May want a defensive check that the length here is <= 32, which is may not be the case with certain inputs, but should never occur.

Reply:
I elected to make that check at the [caller](https://github.com/ava-labs/awm-relayer/blob/1aca8b40c2181862a7fd70c0fd1d5d6353263d18/vms/evm/signer/kms_signer.go#L129) instead, that way the call to `adjustSignatureLength` could still be made inline.